pkgs/by-name/ko/koboldcpp/package.nix

```diff
@@ -133,10 +149,26 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     runHook postInstall
   '';
 
+  # `env.NIX_CFLAGS_COMPILE` is not respected by KoboldCpp, so the Makefile has to be manually patched
```

Comment:
`NIX_CFLAGS_COMPILE` is always respected because it's interpreted by the cc-wrapper; the reason the flags didn't have any effect is because the Makefile adds flags, and the cc-wrapper adds the same flags with different values, and these are all passed on to gcc which is free to choose which ones to respect

By removing the conflicting flags from the Makefile entirely, we delegate the configuration job to the cc-wrapper and all that smart nixpgks tooling which I mostly don't understand

Reply:
Thanks for the clarification but at the same time in `Makefile` there is only this part about `-mcpu`

https://github.com/LostRuins/koboldcpp/blob/cc133401db7d3271ff3907d7c79e4bf99d25ae83/Makefile#L302-L308

Which cannot get triggered on any Darwin machine yet the `-mcpu` flag gets ignored in `NIX_CFLAGS_COMPILE`, so I'm not too sure just removing stuff from `Makefile` will do the trick, at least on macOS since it uses `clang` to build

Reply:
H'm. Maybe we step back: how do you know `mcpu` is ignored?

> it uses clang to build

Clang is used through the cc-wrapper as well

Reply:
I'm not on the PC anymore *(I will edit this comment later today with a log)* but it will absent in the CFLAGS print at the start of the build.

Reply:
Here is the log of `koboldcpp` build on `aarch64-darwin`:

```console
Running phase: buildPhase
build flags: -j8 SHELL=/nix/store/s30jrpgav677fpc9yvkqsib70xfmx7xi-bash-5.2p26/bin/bash LLAMA_VULKAN=1 LLAMA_METAL=1
I llama.cpp build info:
I UNAME_S:  Darwin
I UNAME_P:  arm
I UNAME_M:  arm64
I CFLAGS:   -I. -Iggml/include -Iggml/src -Iinclude -Isrc -I./include -I./include/CL -I./otherarch -I./otherarch/tools -I./otherarch/sdcpp -I./otherarch/sdcpp/thirdparty -I./include/vulkan -O3 -fno-finite-math-only -DNDEBUG -std=c11   -fPIC -DLOG_DISABLE_LOGS -D_GNU_SOURCE -DGGML_USE_LLAMAFILE -pthread -s -Wno-deprecated -Wno-deprecated-declarations -pthread -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DGGML_USE_METAL -DGGML_METAL_NDEBUG -DSD_USE_METAL
I CXXFLAGS: -I. -Iggml/include -Iggml/src -Iinclude -Isrc -I./common -I./include -I./include/CL -I./otherarch -I./otherarch/tools -I./otherarch/sdcpp -I./otherarch/sdcpp/thirdparty -I./include/vulkan -O3 -fno-finite-math-only -DNDEBUG -std=c++11 -fPIC -DLOG_DISABLE_LOGS -D_GNU_SOURCE -DGGML_USE_LLAMAFILE -pthread -s -Wno-multichar -Wno-write-strings -Wno-deprecated -Wno-deprecated-declarations -pthread -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DGGML_USE_METAL -DSD_USE_METAL
I LDFLAGS:  -framework Accelerate -framework Foundation -framework Metal -framework MetalKit -framework MetalPerformanceShaders
I CC:       clang version 16.0.6
I CXX:      clang version 16.0.6
```

This is when I'm using `env.NIX_CFLAGS_COMPILE` without `postPatch`. I had the mcpu set to `apple-m1`, but as you can see, it's absent.

Reply:
> `EDIT: Could this have been the effect of LLAMA_NATIVE=ON which sets march=native?`

Hmm? That's in the `CMakeLists.txt` we're using `Makefile`

Reply:
>  I had the mcpu set to apple-m1, but as you can see, it's absent.

Oh, these must just be the flags passed by the Makefile. The makefile passes these flags to cc-wrapper, and the cc-wrapper passes them to gcc/clang together with NIX_CFLAGS_COMPILE. I suppose you could inspect the final binary to see what kind of instructions it uses

Reply:
After setting the `env.NIX_DEBUG = 1;` I was able to spot the `apple-m1` flags. I suppose in this case, I will implement the `postPatch` that just simply deletes the `march`, `mtune`, and `mcpu` so they can be defined by Nix